### PR TITLE
[10.0][ADD] Add delivery service

### PR DESCRIPTION
**Pull request by acsonefho**
Add a delivery service.
The goal of this service is only to return `stock.picking` related to the connected user (only OUT pickings).
If the connected user is the anonymous one, it should return anything!

### CODE REVIEW

File: shopinvader_delivery_carrier/services/delivery.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import fields
+from odoo.addons.base_rest.components.service import to_int
+from odoo.addons.component.core import Component
+
+
+class AbstractSaleService(Component):
```
Comment (lmignon) [inline shopinvader_delivery_carrier/services/delivery.py]: @acsonefho Your class name should be `ShopinvaderDeliveryService`

File: shopinvader_delivery_carrier/services/delivery.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            "per_page": {
+                "coerce": to_int,
+                "nullable": True,
+                "type": "integer",
+            },
+            "page": {"coerce": to_int, "nullable": True, "type": "integer"},
+        }
+        return schema
+
+    def _get_parser_picking(self):
```
Comment (lmignon) [inline shopinvader_delivery_carrier/services/delivery.py:82]: @acsonefho Why not an ir.export instance? If we start to 'hard code' the returned result it becomes impossible to add a new field into the result without modifying the code....

File: shopinvader_delivery_carrier/services/delivery.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        :return: list
+        """
+        to_parse = [
+            "id:delivery_id",
+            "carrier_tracking_ref:tracking_reference",
+            ("carrier_id:carrier", ("name",)),
+        ]
+        return to_parse
+
+    def _get_parser_sale_order(self):
```
Comment (lmignon) [inline shopinvader_delivery_carrier/services/delivery.py:94]: @acsonefho For new services we should start to provide the schema for the responses..

**Review by lmignon — CHANGES_REQUESTED** [review]

File: shopinvader_delivery_carrier/services/delivery.py (hunk 2 above)
Comment (acsonefho) [reply shopinvader_delivery_carrier/services/delivery.py:82]: @lmignon I don't understand why I should add this? Into shopinvader modules, I don't see any module who is using an `ir.exports` instance as a parser.
And to have the related `ir.exports`, I should search on the `resource = 'stock.picking'` but if another module/service add a new exporter (who doesn't match with mine), how I can manage it?
Thanks

**Comment by Cedric-Pigeon** [comment]
@acsonefho We need to be more structured by providing a dedicated service per feature. Like @lmignon said here: All the functionalities related to delivery should be provided by dedicated services. A first step has already been implemented for version 9.0 https://github.com/shopinvader/odoo-shopinvader-carrier/pull/1#issuecomment-517946707 and provides a more modular approach and a cleaner code.

File: shopinvader_delivery_carrier/services/delivery.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        """
+        Get every stock.picking OUT related to current user.
+        If the current user is the anonymous one, it'll return an invalid
+        domain (to have 0 picking as result)
+        :return:
+        """
+        if self.shopinvader_backend.anonymous_partner_id == self.partner:
+            return [(0, "=", 1)]
+        return [
+            ("partner_id", "=", self.partner.id),
```
Comment (lmignon) [inline shopinvader_delivery_carrier/services/delivery.py]: @acsonefho The service should only returns deliveries linked to a SO linked to the backend.

**Review by lmignon — CHANGES_REQUESTED** [review]
Near the finish line... one last comment

File: shopinvader_delivery_carrier/services/delivery.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import fields
+from odoo.addons.base_rest.components.service import to_int
+from odoo.addons.component.core import Component
+
+
+class AbstractSaleService(Component):
```
Comment (lmignon) [inline shopinvader_delivery_carrier/services/delivery.py]: `s/AbstractSaleService/DeliveryService/`

**Comment by acsonefho** [comment]
@lmignon I think this one is ready now :)
Thank you for your review and your help!

**Review by lmignon — APPROVED** [review]

**Review by Cedric-Pigeon — APPROVED** [review]

**Pull request merged**